### ISSUE TITLE
Gloss outputs Invalid JSON?

### ISSUE THREAD

**Issue by AdamLenda** [hkellaway/Gloss#295]
# XCTAssert-Am-I-An-Idiot()
I built the following test case to see if I could get valid JSON output by using your example code from the documentation.

Please run the following test. My webserver is running PHP 7.2 and using json_decode is throwing errors about invalid JSON. So took the Gloss output from my logs and ran it through a JSON validator. I happened to use this [one](https://jsonformatter.curiousconcept.com/)

So there has to be something I'm just not doing correctly right? 

Please take a look at this test and see if I'm missing something. I really hope I'm just being dumb. Because other wise I just waisted 4 hours trying to get this to work correctly for nothing :D
# Test Case...
```
import Foundation

import XCTest
import Gloss

struct RepoOwner: Decodable, Encodable {
    
    let ownerId: Int
    let username: String?
    
    init?(json: JSON) {
        guard let ownerId: Int = "id" <~~ json else {
            return nil
        }
        
        self.ownerId = ownerId
        self.username = "login" <~~ json
    }
    
    func toJSON() -> JSON?
    {
        return jsonify([
            "id" ~~> self.ownerId,
            "login" ~~> self.username,
            ])
    }
}

struct Repo: Decodable, Encodable {
    
    let repoId: Int?
    let name: String?
    let desc: String?
    let url: NSURL?
    let owner: RepoOwner?
    let primaryLanguage: Language?
    
    enum Language: String {
        case Swift = "Swift"
        case ObjectiveC = "Objective-C"
    }
    
    init?(json: JSON) {
        self.repoId = "id" <~~ json
        self.name = "name" <~~ json
        self.desc = "description" <~~ json
        self.url = "html_url" <~~ json
        self.owner = "owner" <~~ json
        self.primaryLanguage = "language" <~~ json
    }
    
    func toJSON() -> JSON?
    {
        return jsonify([
            "id" ~~> self.repoId,
            "name" ~~> self.name,
            "description" ~~> self.desc,
            "html_url" ~~> self.url,
            "owner" ~~> self.owner,
            "language" ~~> self.primaryLanguage,
            ], keyPathDelimiter:)
    }
    
}

class GlossIsProducingInvalidJSON: XCTestCase {
    override func setUp() {
        super.setUp()
    }
    
    override func tearDown() {
        // Put teardown code here. This method is called after the invocation of each test method in the class.
        super.tearDown()
    }

    func testThatItIsTrue() {
        //valid JS as per https://jsonformatter.curiousconcept.com/
        let jsonString = "{\"id\" : 40102424,\"name\": \"Gloss\", \"description\" : \"A shiny JSON parsing library in Swift\", \"html_url\" : \"https://github.com/hkellaway/Gloss\", \"owner\" : { \"id\" : 5456481, \"login\" : \"hkellaway\" }, \"language\" : \"Swift\" }"
        print(jsonString)

        var json: Any
        do {
            json = try JSONSerialization.jsonObject(with: jsonString.data(using: .utf8)!)
        } catch {
            XCTFail("JSON Deserialization Failed")
            return;
        }
        
        guard let repo = Repo(json: json as! JSON) else {
            XCTFail("Repo Decode Failed")
            return;
        }
        print("\n\n\n")
        print(repo)
        print("\n\n\n")
        print("The following output does IS NOT VALID JSON and cannot be consumed by json_decode in PHP\n")
        debugPrint(repo.toJSON()!)
        print("Check validity here: https://jsonformatter.curiousconcept.com/")
       
        print("The following output DOES OUTPUT VALID JSON\n")
        let jsonData = try! JSONSerialization.data(withJSONObject: repo.toJSON()!)
        print(String.init(data: jsonData, encoding: .utf8)!)
        
        print("Check validity here: https://jsonformatter.curiousconcept.com/")
        print("\n\n\n")
        
        print("Success")
    }
}
```

# Test Output...
```
Test Case '-[NCommon_iOSTests.GlossIsShitTest testThatItIsTrue]' started.
{"id" : 40102424,"name": "Gloss", "description" : "A shiny JSON parsing library in Swift", "html_url" : "https://github.com/hkellaway/Gloss", "owner" : { "id" : 5456481, "login" : "hkellaway" }, "language" : "Swift" }




Repo(repoId: Optional(40102424), name: Optional("Gloss"), desc: Optional("A shiny JSON parsing library in Swift"), url: nil, owner: Optional(NCommon_iOSTests.RepoOwner(ownerId: 5456481, username: Optional("hkellaway"))), primaryLanguage: Optional(NCommon_iOSTests.Repo.Language.Swift))




The following output does IS NOT VALID JSON

["description": "A shiny JSON parsing library in Swift", "name": "Gloss", "owner": ["id": 5456481, "login": "hkellaway"], "language": "Swift", "id": 40102424]


The following output does OUTPUT VALID JSON

{"description":"A shiny JSON parsing library in Swift","name":"Gloss","owner":{"id":5456481,"login":"hkellaway"},"language":"Swift","id":40102424}
Check validity here: https://jsonformatter.curiousconcept.com/



Success
Test Case '-[NCommon_iOSTests.GlossIsShitTest testThatItIsTrue]' passed (0.004 seconds).
```

**Comment by hkellaway** [comment]
I've not run your test, just read through this. what is invalid about this JSON?

["description": "A shiny JSON parsing library in Swift", "name": "Gloss", "owner": ["id": 5456481, "login": "hkellaway"], "language": "Swift", "id": 40102424]

**Comment by AdamLenda** [comment]
I've updated the example code. I found the problem to be that I was trying to output the result of the .toJSON call.

```
        print("Doing it wrong....")
        print(repo.toJSON()!)
       
        print("Doing it right.....")
        let jsonData = try! JSONSerialization.data(withJSONObject: repo.toJSON()!)
        print(jsonData)
```

I think that if you updated the documentation to include the example code applying the library in practice it would help. I am new to Swift and didn't realize that .toJSON was supposed to be filtered through another function before it would actually be valid json.

Thank you for your time and quick reply.
